pkg/sql/schemachanger/sctest/test_server_factory.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sctest
+
+import (
+	"context"
+	gosql "database/sql"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scexec"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// TestServerFactory constructs test clusters for declarative schema changer
+// end-to-end tests.
+type TestServerFactory interface {
+
+	// WithSchemaChangerKnobs returns a copy of this factory but which
+	// will install the provided declarative schema changer testing knobs.
+	WithSchemaChangerKnobs(knobs *scexec.TestingKnobs) TestServerFactory
+
+	// WithMixedVersion returns a copy of this factory but which will
+	// create a cluster in a mixed-version state. The active version
+	// will be the minimum supported binary version.
+	WithMixedVersion() TestServerFactory
+
+	// Run creates a test cluster and applies fn to it.
+	Run(
+		ctx context.Context,
+		t *testing.T,
+		fn func(s serverutils.TestServerInterface, tdb *gosql.DB),
+	)
+}
```

Comment:
`TestServerFactory` and its vanilla and multi-region implementations replace the old `newClusterFunc`.